File: src/wuyu/codec.py
"""JSONL codec for the Codex App Server protocol.

The wire format is newline-delimited JSON (JSONL/NDJSON). Each line is one
complete JSON-RPC message. No HTTP headers, no Content-Length framing.

Encoding: model → JSON string + newline
Decoding: JSON string → discriminated JSONRPCMessage variant

Discrimination logic (since there's no "jsonrpc" field):
  - Has "id" + "method"  → Request
  - Has "method" only    → Notification
  - Has "id" + "result"  → Response
  - Has "id" + "error"   → ErrorResponse
"""

from __future__ import annotations

import json
from typing import Any

from wuyu.protocol.jsonrpc import (
    JSONRPCError,
    JSONRPCErrorResponse,
    JSONRPCMessage,
    JSONRPCNotification,
    JSONRPCRequest,
    JSONRPCResponse,
)


class CodecError(Exception):
    """Raised when a message cannot be encoded or decoded."""


def encode(message: JSONRPCMessage) -> str:
    """Serialize a JSON-RPC message to a JSONL line (with trailing newline).

    Uses by_alias=True so pydantic models with camelCase aliases serialize
    correctly, and exclude_none=True to omit null optional fields.
    """
    if hasattr(message, "model_dump"):
        data = message.model_dump(by_alias=True, exclude_none=True)
    else:
        data = message  # type: ignore[assignment]
    return json.dumps(data, separators=(",", ":")) + "\n"


def decode(line: str) -> JSONRPCMessage:
    """Parse a single JSONL line into a JSONRPCMessage variant.

    Raises CodecError if the line is not valid JSON or doesn't match any variant.
    """
    line = line.strip()
    if not line:
        raise CodecError("empty line")

    try:
        data: dict[str, Any] = json.loads(line)
    except json.JSONDecodeError as e:
        raise CodecError(f"invalid JSON: {e}") from e

    if not isinstance(data, dict):
        raise CodecError(f"expected JSON object, got {type(data).__name__}")

    return _classify(data)


def _classify(data: dict[str, Any]) -> JSONRPCMessage:
    """Determine the message variant from its fields."""
    has_id = "id" in data
    has_method = "method" in data
    has_result = "result" in data
    has_error = "error" in data

    if has_id and has_error:
        return JSONRPCErrorResponse(
            id=data["id"],
            error=JSONRPCError.model_validate(data["error"]),
        )

    if has_id and has_result:
        return JSONRPCResponse(id=data["id"], result=data.get("result"))

    if has_id and has_method:
        return JSONRPCRequest(
            id=data["id"],
            method=data["method"],
            params=data.get("params"),
        )

    if has_method and not has_id:
        return JSONRPCNotification(
            method=data["method"],
            params=data.get("params"),
        )

    # Response with just id and no result/error — treat as empty result
    if has_id and not has_method and not has_error:
        return JSONRPCResponse(id=data["id"], result=data.get("result"))

    raise CodecError(f"cannot classify message: {data!r}")

File: wuyu_dart/lib/src/codec.dart
/// JSONL codec for the Codex App Server protocol.
///
/// Wire format: newline-delimited JSON, one object per line.
/// No HTTP headers, no Content-Length framing, no "jsonrpc":"2.0" field.
///
/// Encoding: message → compact JSON + '\n'
/// Decoding: JSONL line → discriminated message variant
library;

import 'dart:convert';

import 'protocol/jsonrpc.dart';

export 'protocol/jsonrpc.dart';

/// Thrown when a message cannot be encoded or decoded.
final class CodecError implements Exception {
  final String message;
  const CodecError(this.message);

  @override
  String toString() => 'CodecError: $message';
}

/// Serialize a JSON-RPC message to a JSONL line (with trailing newline).
/// Omits null optional fields (params, result).
String encode(Object message) {
  final Map<String, Object?> data;

  if (message is JsonRpcRequest) {
    data = {
      'id': message.id,
      'method': message.method,
      if (message.params != null) 'params': message.params,
    };
  } else if (message is JsonRpcNotification) {
    data = {
      'method': message.method,
      if (message.params != null) 'params': message.params,
    };
  } else if (message is JsonRpcResponse) {
    data = {
      'id': message.id,
      if (message.result != null) 'result': message.result,
    };
  } else if (message is JsonRpcErrorResponse) {
    data = {
      'id': message.id,
      'error': {
        'code': message.error.code,
        'message': message.error.message,
        if (message.error.data != null) 'data': message.error.data,
      },
    };
  } else {
    throw CodecError('unknown message type: ${message.runtimeType}');
  }

  return '${jsonEncode(data)}\n';
}

/// Parse a single JSONL line into a message variant.
/// Throws [CodecError] if the line is not valid JSON or cannot be classified.
Object decode(String line) {
  line = line.trim();
  if (line.isEmpty) throw const CodecError('empty line');

  final Object? parsed;
  try {
    parsed = jsonDecode(line);
  } on FormatException catch (e) {
    throw CodecError('invalid JSON: $e');
  }

  if (parsed is! Map) {
    throw CodecError('expected JSON object, got ${parsed.runtimeType}');
  }

  final data = parsed.cast<String, Object?>();
  return _classify(data);
}

Object _classify(Map<String, Object?> data) {
  final hasId = data.containsKey('id');
  final hasMethod = data.containsKey('method');
  final hasResult = data.containsKey('result');
  final hasError = data.containsKey('error');

  if (hasId && hasError) {
    final e = (data['error'] as Map).cast<String, Object?>();
    return JsonRpcErrorResponse(
      id: data['id']!,
      error: JsonRpcError(
        code: e['code'] as int,
        message: e['message'] as String,
        data: e['data'],
      ),
    );
  }

  if (hasId && hasResult) {
    return JsonRpcResponse(id: data['id']!, result: data['result']);
  }

  if (hasId && hasMethod) {
    final params = data['params'];
    return JsonRpcRequest(
      id: data['id']!,
      method: data['method'] as String,
      params: params != null ? (params as Map).cast<String, Object?>() : null,
    );
  }

  if (hasMethod && !hasId) {
    final params = data['params'];
    return JsonRpcNotification(
      method: data['method'] as String,
      params: params != null ? (params as Map).cast<String, Object?>() : null,
    );
  }

  // id with no method/result/error — treat as null result response
  if (hasId) {
    return JsonRpcResponse(id: data['id']!);
  }

  throw CodecError('cannot classify message: $data');
}
